// @mgonetwork/mango.js/dist/cjs/types/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var types_exports = {};
__export(types_exports, {
  AddressMetrics: () => import_metrics.AddressMetrics,
  AllEpochsAddressMetrics: () => import_metrics.AllEpochsAddressMetrics,
  Apy: () => import_validator.Apy,
  AuthorityName: () => import_transactions.AuthorityName,
  AuthoritySignature: () => import_transactions.AuthoritySignature,
  Balance: () => import_validator.Balance,
  BalanceChange: () => import_transactions.BalanceChange,
  CheckPointContentsDigest: () => import_checkpoints.CheckPointContentsDigest,
  Checkpoint: () => import_checkpoints.Checkpoint,
  CheckpointCommitment: () => import_checkpoints.CheckpointCommitment,
  CheckpointDigest: () => import_checkpoints.CheckpointDigest,
  CheckpointPage: () => import_checkpoints.CheckpointPage,
  CheckpointedObjectId: () => import_objects.CheckpointedObjectId,
  CoinBalance: () => import_coin.CoinBalance,
  CoinStruct: () => import_coin.CoinStruct,
  CoinSupply: () => import_coin.CoinSupply,
  CommitteeInfo: () => import_validator.CommitteeInfo,
  Contents: () => import_validator2.Contents,
  ContentsFields: () => import_validator2.ContentsFields,
  ContentsFieldsWithdraw: () => import_validator2.ContentsFieldsWithdraw,
  DelegatedStake: () => import_validator.DelegatedStake,
  DelegationStakingPool: () => import_validator2.DelegationStakingPool,
  DelegationStakingPoolFields: () => import_validator2.DelegationStakingPoolFields,
  DevInspectResults: () => import_transactions.DevInspectResults,
  DisplayFieldsBackwardCompatibleResponse: () => import_objects.DisplayFieldsBackwardCompatibleResponse,
  DisplayFieldsResponse: () => import_objects.DisplayFieldsResponse,
  DryRunTransactionBlockResponse: () => import_transactions.DryRunTransactionBlockResponse,
  DynamicFieldInfo: () => import_dynamic_fields.DynamicFieldInfo,
  DynamicFieldName: () => import_dynamic_fields.DynamicFieldName,
  DynamicFieldPage: () => import_dynamic_fields.DynamicFieldPage,
  DynamicFieldType: () => import_dynamic_fields.DynamicFieldType,
  EndOfEpochData: () => import_checkpoints.EndOfEpochData,
  EndOfEpochInfo: () => import_epochs.EndOfEpochInfo,
  EpochId: () => import_transactions.EpochId,
  EpochInfo: () => import_epochs.EpochInfo,
  EpochPage: () => import_epochs.EpochPage,
  EventId: () => import_events.EventId,
  ExecutionStatus: () => import_transactions.ExecutionStatus,
  ExecutionStatusType: () => import_transactions.ExecutionStatusType,
  GasCostSummary: () => import_checkpoints.GasCostSummary,
  Genesis: () => import_transactions.Genesis,
  GetOwnedObjectsResponse: () => import_objects.GetOwnedObjectsResponse,
  MgoAddress: () => import_common.MgoAddress,
  MgoArgument: () => import_transactions.MgoArgument,
  MgoCallArg: () => import_transactions.MgoCallArg,
  MgoChangeEpoch: () => import_transactions.MgoChangeEpoch,
  MgoConsensusCommitPrologue: () => import_transactions.MgoConsensusCommitPrologue,
  MgoEvent: () => import_events.MgoEvent,
  MgoGasData: () => import_objects.MgoGasData,
  MgoJsonValue: () => import_common.MgoJsonValue,
  MgoMoveAbilitySet: () => import_normalized.MgoMoveAbilitySet,
  MgoMoveFunctionArgType: () => import_normalized.MgoMoveFunctionArgType,
  MgoMoveFunctionArgTypes: () => import_normalized.MgoMoveFunctionArgTypes,
  MgoMoveModuleId: () => import_normalized.MgoMoveModuleId,
  MgoMoveNormalizedField: () => import_normalized.MgoMoveNormalizedField,
  MgoMoveNormalizedFunction: () => import_normalized.MgoMoveNormalizedFunction,
  MgoMoveNormalizedModule: () => import_normalized.MgoMoveNormalizedModule,
  MgoMoveNormalizedModules: () => import_normalized.MgoMoveNormalizedModules,
  MgoMoveNormalizedStruct: () => import_normalized.MgoMoveNormalizedStruct,
  MgoMoveNormalizedStructType: () => import_normalized.MgoMoveNormalizedStructType,
  MgoMoveNormalizedType: () => import_normalized.MgoMoveNormalizedType,
  MgoMoveNormalizedTypeParameterType: () => import_normalized.MgoMoveNormalizedTypeParameterType,
  MgoMoveObject: () => import_objects.MgoMoveObject,
  MgoMovePackage: () => import_objects.MgoMovePackage,
  MgoMoveStructTypeParameter: () => import_normalized.MgoMoveStructTypeParameter,
  MgoMoveVisibility: () => import_normalized.MgoMoveVisibility,
  MgoObjectChange: () => import_transactions.MgoObjectChange,
  MgoObjectChangeCreated: () => import_transactions.MgoObjectChangeCreated,
  MgoObjectChangeDeleted: () => import_transactions.MgoObjectChangeDeleted,
  MgoObjectChangeMutated: () => import_transactions.MgoObjectChangeMutated,
  MgoObjectChangePublished: () => import_transactions.MgoObjectChangePublished,
  MgoObjectChangeTransferred: () => import_transactions.MgoObjectChangeTransferred,
  MgoObjectChangeWrapped: () => import_transactions.MgoObjectChangeWrapped,
  MgoObjectData: () => import_objects.MgoObjectData,
  MgoObjectDataOptions: () => import_objects.MgoObjectDataOptions,
  MgoObjectInfo: () => import_objects.MgoObjectInfo,
  MgoObjectRef: () => import_objects.MgoObjectRef,
  MgoObjectResponse: () => import_objects.MgoObjectResponse,
  MgoObjectResponseError: () => import_objects.MgoObjectResponseError,
  MgoParsedData: () => import_objects.MgoParsedData,
  MgoRawData: () => import_objects.MgoRawData,
  MgoRawMoveObject: () => import_objects.MgoRawMoveObject,
  MgoRawMovePackage: () => import_objects.MgoRawMovePackage,
  MgoSupplyFields: () => import_validator2.MgoSupplyFields,
  MgoSystemStateSummary: () => import_validator.MgoSystemStateSummary,
  MgoTransaction: () => import_transactions.MgoTransaction,
  MgoTransactionBlock: () => import_transactions.MgoTransactionBlock,
  MgoTransactionBlockData: () => import_transactions.MgoTransactionBlockData,
  MgoTransactionBlockResponse: () => import_transactions.MgoTransactionBlockResponse,
  MgoTransactionBlockResponseOptions: () => import_transactions.MgoTransactionBlockResponseOptions,
  MgoValidatorSummary: () => import_validator.MgoValidatorSummary,
  MoveCallMetric: () => import_normalized.MoveCallMetric,
  MoveCallMetrics: () => import_normalized.MoveCallMetrics,
  MoveCallMgoTransaction: () => import_transactions.MoveCallMgoTransaction,
  MovePackageContent: () => import_objects.MovePackageContent,
  NetworkMetrics: () => import_metrics.NetworkMetrics,
  ObjectContentFields: () => import_objects.ObjectContentFields,
  ObjectDigest: () => import_objects.ObjectDigest,
  ObjectId: () => import_common.ObjectId,
  ObjectOwner: () => import_common.ObjectOwner,
  ObjectRead: () => import_objects.ObjectRead,
  ObjectStatus: () => import_objects.ObjectStatus,
  ObjectType: () => import_objects.ObjectType,
  OwnedObjectRef: () => import_transactions.OwnedObjectRef,
  PaginatedCoins: () => import_coin.PaginatedCoins,
  PaginatedEvents: () => import_events.PaginatedEvents,
  PaginatedObjectsResponse: () => import_objects.PaginatedObjectsResponse,
  PaginatedTransactionResponse: () => import_transactions.PaginatedTransactionResponse,
  ProgrammableTransaction: () => import_transactions.ProgrammableTransaction,
  ProtocolConfig: () => import_common.ProtocolConfig,
  ResolvedNameServiceNames: () => import_name_service.ResolvedNameServiceNames,
  SequenceNumber: () => import_common.SequenceNumber,
  StakeObject: () => import_validator.StakeObject,
  StakeSubsidy: () => import_validator2.StakeSubsidy,
  StakeSubsidyFields: () => import_validator2.StakeSubsidyFields,
  TransactionDigest: () => import_common.TransactionDigest,
  TransactionEffects: () => import_transactions.TransactionEffects,
  TransactionEffectsDigest: () => import_common.TransactionEffectsDigest,
  TransactionEffectsModifiedAtVersions: () => import_transactions.TransactionEffectsModifiedAtVersions,
  TransactionEventDigest: () => import_common.TransactionEventDigest,
  TransactionEvents: () => import_transactions.TransactionEvents,
  ValidatorSignature: () => import_checkpoints.ValidatorSignature,
  Validators: () => import_validator.Validators,
  ValidatorsApy: () => import_validator.ValidatorsApy,
  extractMutableReference: () => import_normalized.extractMutableReference,
  extractReference: () => import_normalized.extractReference,
  extractStructTag: () => import_normalized.extractStructTag,
  getChangeEpochTransaction: () => import_transactions.getChangeEpochTransaction,
  getConsensusCommitPrologueTransaction: () => import_transactions.getConsensusCommitPrologueTransaction,
  getCreatedObjects: () => import_transactions.getCreatedObjects,
  getEventPackage: () => import_events.getEventPackage,
  getEventSender: () => import_events.getEventSender,
  getEvents: () => import_transactions.getEvents,
  getExecutionStatus: () => import_transactions.getExecutionStatus,
  getExecutionStatusError: () => import_transactions.getExecutionStatusError,
  getExecutionStatusGasSummary: () => import_transactions.getExecutionStatusGasSummary,
  getExecutionStatusType: () => import_transactions.getExecutionStatusType,
  getGasData: () => import_transactions.getGasData,
  getMgoObjectData: () => import_objects.getMgoObjectData,
  getMoveObject: () => import_objects.getMoveObject,
  getMoveObjectType: () => import_objects.getMoveObjectType,
  getMovePackageContent: () => import_objects.getMovePackageContent,
  getNewlyCreatedCoinRefsAfterSplit: () => import_transactions.getNewlyCreatedCoinRefsAfterSplit,
  getObjectChanges: () => import_transactions.getObjectChanges,
  getObjectDeletedResponse: () => import_objects.getObjectDeletedResponse,
  getObjectDisplay: () => import_objects.getObjectDisplay,
  getObjectFields: () => import_objects.getObjectFields,
  getObjectId: () => import_objects.getObjectId,
  getObjectNotExistsResponse: () => import_objects.getObjectNotExistsResponse,
  getObjectOwner: () => import_objects.getObjectOwner,
  getObjectPreviousTransactionDigest: () => import_objects.getObjectPreviousTransactionDigest,
  getObjectReference: () => import_objects.getObjectReference,
  getObjectType: () => import_objects.getObjectType,
  getObjectVersion: () => import_objects.getObjectVersion,
  getProgrammableTransaction: () => import_transactions.getProgrammableTransaction,
  getPublishedObjectChanges: () => import_transactions.getPublishedObjectChanges,
  getSharedObjectInitialVersion: () => import_objects.getSharedObjectInitialVersion,
  getTimestampFromTransactionResponse: () => import_transactions.getTimestampFromTransactionResponse,
  getTotalGasUsed: () => import_transactions.getTotalGasUsed,
  getTotalGasUsedUpperBound: () => import_transactions.getTotalGasUsedUpperBound,
  getTransaction: () => import_transactions.getTransaction,
  getTransactionDigest: () => import_transactions.getTransactionDigest,
  getTransactionEffects: () => import_transactions.getTransactionEffects,
  getTransactionGasBudget: () => import_transactions.getTransactionGasBudget,
  getTransactionGasObject: () => import_transactions.getTransactionGasObject,
  getTransactionGasPrice: () => import_transactions.getTransactionGasPrice,
  getTransactionKind: () => import_transactions.getTransactionKind,
  getTransactionKindName: () => import_transactions.getTransactionKindName,
  getTransactionSender: () => import_transactions.getTransactionSender,
  getTransactionSignature: () => import_transactions.getTransactionSignature,
  hasPublicTransfer: () => import_objects.hasPublicTransfer,
  isImmutableObject: () => import_objects.isImmutableObject,
  isMgoObjectResponse: () => import_objects.isMgoObjectResponse,
  isSharedObject: () => import_objects.isSharedObject
});
module.exports = __toCommonJS(types_exports);
var import_common = require("./common.js");
var import_objects = require("./objects.js");
var import_events = require("./events.js");
var import_transactions = require("./transactions.js");
var import_normalized = require("./normalized.js");
var import_validator = require("./validator.js");
var import_coin = require("./coin.js");
var import_epochs = require("./epochs.js");
var import_name_service = require("./name-service.js");
var import_dynamic_fields = require("./dynamic_fields.js");
var import_checkpoints = require("./checkpoints.js");
var import_metrics = require("./metrics.js");
var import_validator2 = require("./validator.js");
