// @mgonetwork/mango.js/dist/esm/index.js
import { fromExportedKeypair } from "./cryptography/utils.js";
import { ECMHLiveObjectSetDigest, ExecutionDigests } from "./types/checkpoints.js";
import { MANGO_PER_MGO, MGO_DECIMALS } from "./types/objects.js";
import {
  AuthorityQuorumSignInfo,
  GenericAuthoritySignature,
  MgoTransactionBlockKind
} from "./types/transactions.js";
export * from "./types/index.js";
import {
  Ed25519Keypair,
  Ed25519PublicKey
} from "./keypairs/ed25519/index.js";
import {
  DEFAULT_SECP256K1_DERIVATION_PATH,
  Secp256k1Keypair,
  Secp256k1PublicKey
} from "./keypairs/secp256k1/index.js";
import {
  DEFAULT_SECP256R1_DERIVATION_PATH,
  Secp256r1Keypair,
  Secp256r1PublicKey
} from "./keypairs/secp256r1/index.js";
import {
  PRIVATE_KEY_SIZE,
  LEGACY_PRIVATE_KEY_SIZE,
  MGO_PRIVATE_KEY_PREFIX,
  BaseSigner,
  Keypair,
  decodeMgoPrivateKey,
  encodeMgoPrivateKey
} from "./cryptography/keypair.js";
import {
  MAX_SIGNER_IN_MULTISIG,
  combinePartialSigs,
  decodeMultiSig,
  toMultiSigAddress
} from "./cryptography/multisig.js";
import { PublicKey, bytesEqual } from "./cryptography/publickey.js";
import {
  isValidBIP32Path,
  isValidHardenedPath,
  mnemonicToSeed,
  mnemonicToSeedHex
} from "./cryptography/mnemonics.js";
import {
  parseSerializedSignature,
  toSerializedSignature
} from "./cryptography/signature.js";
import {
  SIGNATURE_FLAG_TO_SCHEME,
  SIGNATURE_SCHEME_TO_FLAG,
  SIGNATURE_SCHEME_TO_SIZE
} from "./cryptography/signature-scheme.js";
import {
  publicKeyFromSerialized,
  toParsedSignaturePubkeyPair,
  toSingleSignaturePubkeyPair
} from "./cryptography/utils.js";
import {
  JsonRpcProvider
} from "./providers/json-rpc-provider.js";
import { JsonRpcClient } from "./rpc/client.js";
import {
  Connection,
  devnetConnection,
  localnetConnection,
  mainnetConnection,
  testnetConnection
} from "./rpc/connection.js";
import { RawSigner } from "./signers/raw-signer.js";
import { SignerWithProvider } from "./signers/signer-with-provider.js";
import {
  AppId,
  IntentScope,
  IntentVersion,
  messageWithIntent
} from "./cryptography/intent.js";
import { verifyMessage } from "./utils/verify.js";
import { RPCValidationError } from "./rpc/errors.js";
import { fromB64, toB64 } from "@mgonetwork/bcs";
import {
  MGO_ADDRESS_LENGTH,
  isValidMgoAddress,
  isValidMgoObjectId,
  isValidTransactionDigest,
  normalizeStructTag,
  normalizeMgoAddress,
  normalizeMgoObjectId,
  parseStructTag
} from "./utils/mgo-types.js";
import { formatAddress, formatDigest } from "./utils/format.js";
import { is, assert } from "superstruct";
import {
  DEFAULT_CLIENT_OPTIONS,
  WebsocketClient,
  getWebsocketUrl
} from "./rpc/websocket-client.js";
import {
  builder,
  Transactions,
  Inputs,
  TransactionBlock,
  TransactionArgument,
  ARGUMENT,
  ARGUMENT_INNER,
  BuilderCallArg,
  CALL_ARG,
  COMPRESSED_SIGNATURE,
  ENUM_KIND,
  MULTISIG,
  MULTISIG_PK_MAP,
  MULTISIG_PUBLIC_KEY,
  MakeMoveVecTransaction,
  MergeCoinsTransaction,
  MoveCallTransaction,
  OBJECT_ARG,
  OPTION,
  ObjectCallArg,
  ObjectTransactionArgument,
  PROGRAMMABLE_CALL,
  PROGRAMMABLE_CALL_INNER,
  PROGRAMMABLE_TX_BLOCK,
  PUBLIC_KEY,
  PublishTransaction,
  PureCallArg,
  PureTransactionArgument,
  SplitCoinsTransaction,
  TRANSACTION,
  TRANSACTION_INNER,
  TYPE_TAG,
  TransactionBlockInput,
  TransactionType,
  TransferObjectsTransaction,
  UpgradePolicy,
  UpgradeTransaction,
  VECTOR,
  getIdFromCallArg,
  getPureSerializationType,
  getSharedObjectInput,
  getTransactionType,
  isMutableSharedObjectInput,
  isSharedObjectInput,
  isTxContext
} from "./builder/index.js";
import {
  ADD_STAKE_FUN_NAME,
  ADD_STAKE_LOCKED_COIN_FUN_NAME,
  COIN_TYPE_ARG_REGEX,
  Coin,
  CoinMetadataStruct,
  Delegation,
  ID_STRUCT_NAME,
  OBJECT_MODULE_NAME,
  PAY_JOIN_COIN_FUNC_NAME,
  PAY_MODULE_NAME,
  PAY_SPLIT_COIN_VEC_FUNC_NAME,
  MgoSystemStateUtil,
  UID_STRUCT_NAME,
  VALIDATORS_EVENTS_QUERY,
  WITHDRAW_STAKE_FUN_NAME,
  isObjectDataFull,
  MGO_CLOCK_OBJECT_ID,
  MGO_FRAMEWORK_ADDRESS,
  MGO_SYSTEM_ADDRESS,
  MGO_SYSTEM_MODULE_NAME,
  MGO_SYSTEM_STATE_OBJECT_ID,
  MGO_TYPE_ARG,
  MOVE_STDLIB_ADDRESS
} from "./framework/index.js";
import {
  bcs,
  isPureArg
} from "./bcs/index.js";
export {
  ADD_STAKE_FUN_NAME,
  ADD_STAKE_LOCKED_COIN_FUN_NAME,
  ARGUMENT,
  ARGUMENT_INNER,
  AppId,
  AuthorityQuorumSignInfo,
  BaseSigner,
  BuilderCallArg,
  CALL_ARG,
  COIN_TYPE_ARG_REGEX,
  COMPRESSED_SIGNATURE,
  Coin,
  CoinMetadataStruct,
  Connection,
  DEFAULT_CLIENT_OPTIONS,
  DEFAULT_SECP256K1_DERIVATION_PATH,
  DEFAULT_SECP256R1_DERIVATION_PATH,
  Delegation,
  ECMHLiveObjectSetDigest,
  ENUM_KIND,
  Ed25519Keypair,
  Ed25519PublicKey,
  ExecutionDigests,
  GenericAuthoritySignature,
  ID_STRUCT_NAME,
  Inputs,
  IntentScope,
  IntentVersion,
  JsonRpcClient,
  JsonRpcProvider,
  Keypair,
  LEGACY_PRIVATE_KEY_SIZE,
  MANGO_PER_MGO,
  MAX_SIGNER_IN_MULTISIG,
  MGO_ADDRESS_LENGTH,
  MGO_CLOCK_OBJECT_ID,
  MGO_DECIMALS,
  MGO_FRAMEWORK_ADDRESS,
  MGO_PRIVATE_KEY_PREFIX,
  MGO_SYSTEM_ADDRESS,
  MGO_SYSTEM_MODULE_NAME,
  MGO_SYSTEM_STATE_OBJECT_ID,
  MGO_TYPE_ARG,
  MOVE_STDLIB_ADDRESS,
  MULTISIG,
  MULTISIG_PK_MAP,
  MULTISIG_PUBLIC_KEY,
  MakeMoveVecTransaction,
  MergeCoinsTransaction,
  MgoSystemStateUtil,
  MgoTransactionBlockKind,
  MoveCallTransaction,
  OBJECT_ARG,
  OBJECT_MODULE_NAME,
  OPTION,
  ObjectCallArg,
  ObjectTransactionArgument,
  PAY_JOIN_COIN_FUNC_NAME,
  PAY_MODULE_NAME,
  PAY_SPLIT_COIN_VEC_FUNC_NAME,
  PRIVATE_KEY_SIZE,
  PROGRAMMABLE_CALL,
  PROGRAMMABLE_CALL_INNER,
  PROGRAMMABLE_TX_BLOCK,
  PUBLIC_KEY,
  PublicKey,
  PublishTransaction,
  PureCallArg,
  PureTransactionArgument,
  RPCValidationError,
  RawSigner,
  SIGNATURE_FLAG_TO_SCHEME,
  SIGNATURE_SCHEME_TO_FLAG,
  SIGNATURE_SCHEME_TO_SIZE,
  Secp256k1Keypair,
  Secp256k1PublicKey,
  Secp256r1Keypair,
  Secp256r1PublicKey,
  SignerWithProvider,
  SplitCoinsTransaction,
  TRANSACTION,
  TRANSACTION_INNER,
  TYPE_TAG,
  TransactionArgument,
  TransactionBlock,
  TransactionBlockInput,
  TransactionType,
  Transactions,
  TransferObjectsTransaction,
  UID_STRUCT_NAME,
  UpgradePolicy,
  UpgradeTransaction,
  VALIDATORS_EVENTS_QUERY,
  VECTOR,
  WITHDRAW_STAKE_FUN_NAME,
  WebsocketClient,
  assert,
  bcs,
  builder,
  bytesEqual,
  combinePartialSigs,
  decodeMgoPrivateKey,
  decodeMultiSig,
  devnetConnection,
  encodeMgoPrivateKey,
  formatAddress,
  formatDigest,
  fromB64,
  fromExportedKeypair,
  getIdFromCallArg,
  getPureSerializationType,
  getSharedObjectInput,
  getTransactionType,
  getWebsocketUrl,
  is,
  isMutableSharedObjectInput,
  isObjectDataFull,
  isPureArg,
  isSharedObjectInput,
  isTxContext,
  isValidBIP32Path,
  isValidHardenedPath,
  isValidMgoAddress,
  isValidMgoObjectId,
  isValidTransactionDigest,
  localnetConnection,
  mainnetConnection,
  messageWithIntent,
  mnemonicToSeed,
  mnemonicToSeedHex,
  normalizeMgoAddress,
  normalizeMgoObjectId,
  normalizeStructTag,
  parseSerializedSignature,
  parseStructTag,
  publicKeyFromSerialized,
  testnetConnection,
  toB64,
  toMultiSigAddress,
  toParsedSignaturePubkeyPair,
  toSerializedSignature,
  toSingleSignaturePubkeyPair,
  verifyMessage
};
