// @mgonetwork/mango.js/dist/cjs/utils/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var utils_exports = {};
__export(utils_exports, {
  MANGO_PER_MGO: () => MANGO_PER_MGO,
  MGO_ADDRESS_LENGTH: () => import_mgo_types.MGO_ADDRESS_LENGTH,
  MGO_CLOCK_OBJECT_ID: () => MGO_CLOCK_OBJECT_ID,
  MGO_DECIMALS: () => MGO_DECIMALS,
  MGO_FRAMEWORK_ADDRESS: () => MGO_FRAMEWORK_ADDRESS,
  MGO_SYSTEM_ADDRESS: () => MGO_SYSTEM_ADDRESS,
  MGO_SYSTEM_MODULE_NAME: () => MGO_SYSTEM_MODULE_NAME,
  MGO_SYSTEM_STATE_OBJECT_ID: () => MGO_SYSTEM_STATE_OBJECT_ID,
  MGO_TYPE_ARG: () => MGO_TYPE_ARG,
  MGO_USDT_ADDRESS: () => MGO_USDT_ADDRESS,
  MOVE_STDLIB_ADDRESS: () => MOVE_STDLIB_ADDRESS,
  assert: () => import_superstruct.assert,
  formatAddress: () => import_format.formatAddress,
  formatDigest: () => import_format.formatDigest,
  fromB64: () => import_bcs.fromB64,
  fromHEX: () => import_bcs.fromHEX,
  is: () => import_superstruct.is,
  isValidMgoAddress: () => import_mgo_types.isValidMgoAddress,
  isValidMgoObjectId: () => import_mgo_types.isValidMgoObjectId,
  isValidTransactionDigest: () => import_mgo_types.isValidTransactionDigest,
  normalizeMgoAddress: () => import_mgo_types.normalizeMgoAddress,
  normalizeMgoObjectId: () => import_mgo_types.normalizeMgoObjectId,
  normalizeStructTag: () => import_mgo_types.normalizeStructTag,
  parseStructTag: () => import_mgo_types.parseStructTag,
  toB64: () => import_bcs.toB64,
  toHEX: () => import_bcs.toHEX
});
module.exports = __toCommonJS(utils_exports);
var import_format = require("./format.js");
var import_mgo_types = require("./mgo-types");
var import_bcs = require("@mgonetwork/bcs");
var import_superstruct = require("superstruct");
const MGO_DECIMALS = 9;
const MANGO_PER_MGO = BigInt(1e9);
const MOVE_STDLIB_ADDRESS = "0x1";
const MGO_FRAMEWORK_ADDRESS = "0x2";
const MGO_SYSTEM_ADDRESS = "0x3";
const MGO_CLOCK_OBJECT_ID = (0, import_mgo_types.normalizeMgoObjectId)("0x6");
const MGO_SYSTEM_MODULE_NAME = "mgo_system";
const MGO_TYPE_ARG = `${MGO_FRAMEWORK_ADDRESS}::mgo::MGO`;
const MGO_SYSTEM_STATE_OBJECT_ID = (0, import_mgo_types.normalizeMgoObjectId)("0x5");
const MGO_USDT_ADDRESS = "0xefd048221d577dfb24218d5ffd03de4f364e2d1bed4e030a861a0ec01cb75b1";
